docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,352 @@
+	// ManagedClusters will be created.
+	ManagedClusterTemplate map[string]interface{} `json:"managedClusterTemplate"`
+}
+```
+
+One variant of this method could be to using `string` instead of `map[string]interface{}` for the template
+type, though that would make defining patches unwieldy (like for ClusterClass).
+
+Compared to [Option 2], this method loses schema and webhook validation that would be performed by ASO when
+creating a separate ManagedCluster to serve as a template. That validation would still be performed when CAPZ
+creates the ManagedCluster resource, but that would be some time after the AzureManagedControlPlane is created
+and error messages may not be quite as visible.
+
+#### Option 4: CAPZ resource defines an entire typed ASO resource inline
+
+This method is functionally equivalent to [Option 3] except that the template field's type is the exact same
+as an ASO ManagedCluster:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+	
+	// ManagedClusterSpec defines the spec of the ASO ManagedCluster managed by this AzureManagedControlPlane.
+	ManagedClusterSpec v1api20230201.ManagedCluster_Spec `json:"managedClusterSpec"`
+}
+```
+
+This method allows CAPZ to leverage schema validation defined by ASO's CRDs upon AzureManagedControlPlane
+creation, but would still lose any further webhook validation done by ASO unless CAPZ can invoke that itself.
+
+It also has the drawback that one version of CAPZ is tied directly to a single AKS API version. The spec could
+potentially contain separate fields for each API version and enforce in webhooks that only one API version is
+being used at a time. Alternatively, users may set fields only present in a newer API version directly on the
+ManagedCluster after creation (if allowed by AKS) because CAPZ will not override user-provided fields for
+which it does not have its own opinion on ASO resources.
+
+While it couples CAPZ to one ASO API version, this approach allows CAPZ to move a more calculated pace with
+regards to AKS API versions the way it's done today. This also narrows CAPZ's scope of responsibility which
+reduces CAPZ's exposure to potential incompatibilities with certain ASO API versions.
+
+Regarding ClusterClass, this option functions the same as [Option 2] or [Option 3], where all ASO fields can
+be defined in a template. This option opens up an additional safeguard though, where webhooks could flag
+fields which should not be defined in a template. Similar webhook checks would be less practical in the other
+options where CAPZ would need to be aware of the set of disallowed fields for each ASO API version that a user
+could use.
+
+Similarly, CAPZ's webhooks are also better able to validate and default the ASO configuration and ensure
+fields like ManagedCluster's `spec.owner` that should not be modified by users are set correctly.
+
+#### Decision
+
+We are opting to move forward with [Option 4]. That approach is most consistent with how the API is defined
```

Comment:
Sorry I missed this. If I understand correctly, we are essentially delegating the job of reconciling entirely to `aso-operator` ?

Reply:
You could say that CAPZ already delegates all reconciliation of AKS resources to ASO. This proposal is primarily exploring how the parameters for those resources should be represented in the CAPZ API.